Allow beeline to match Elixir's OTP naming

pmonson711 opened the pull request:
GenServer's in elixir can be one of several type structures.
https://hexdocs.pm/elixir/GenServer.html#t:name/0

The current implementation of Beeline only works with atoms (which match
to OTP's `{:local, name}` for named processes, but this limits the
places BeeLine can be used as it either leaks atoms or prevents from
having more than one GenServer of the same name running at a single time.
By updating the naming to include the global and via options the user
can decide how to control this instead of the library.

pmonson711 commented:
As a note on the mix.exs updates. This does require `nimble_options >= 0.4.1`, as before that it couldn't handle tuples.

pmonson711 commented on lib/beeline/process_naming.ex:
```diff
@@ -0,0 +1,21 @@
+    {:global, Module.concat(base_name, appended_name)}
+  end
+
+  def name({:via, registry, {registry_name, base_name}}, appended_name) do
```
I just want to call out that this `{:via...}` pattern may not be complete. It will work with elixir's Registry module but it may not work with other common erlang registries like gproc.

the-mikedavis commented:
There was a similar change in Broadway for allowing via tuples: https://github.com/dashbitco/broadway/pull/239. They don't support global tuples but I don't think they're intentionally preventing that.

pmonson711 commented:
It's interesting the approach Broadway takes with the overridable and no default implementation for the name mixed with a guard. That could work for Beeline as well but it seems like it might cause confusion in the config and the existing `:name` field. But I do think I'll add a guard that they added.

pmonson711 commented:
I added a guard and updated the allowable names to match. It's slightly different than before.

Global names now match the OTP type where the second element of the tuple is a term.

Registry names are now limited to Elixir's Registry module, because that's all we test, and matches what Broadway does in their PR.

electricshaman commented on lib/beeline/topology.ex:
```diff
@@ -3,14 +3,13 @@ defmodule Beeline.Topology do
 
   defstruct [:supervisor_pid, :config]
 
   def start_link(config) do
```
If this were `start_link(config, options \\ [])` you could just pass options directly to GenServer.  Then, from another higher level module in your API, write the layer which handles the process name shenanigans.  I feel a soapbox rant coming on.

electricshaman reviewed (approved):
This is my first time ever seeing this library, so I don't have a lot of context built up about its design or usage.  Best to mention that up front.

Your changes seem totally reasonable as an expansion of what's already here.  

Beyond that, my unpopular thoughts: I don't understand the need for all of the ceremony around the naming of processes in the first place.  I think it's admirable to want to hide the complexity of the supervision tree from clients, because honestly it's tedious and can be annoying to deal with, but in reality the people who will be using this library are already likely to know how to spin this up in their own application's supervision tree.  This library seems to jump through a lot of hoops just to keep the Beeline API from exposing the process names and the supervision tree of stages.  There's something to be said for that, but personally I find libraries like that frustrating to work with because they make assumptions about how I want to use OTP in my application and often times they restrict one's ability to troubleshoot when normal options like `name` and `debug: [:trace]` can't be supplied to the start_link function.  </soapbox>

electricshaman commented:
The more I look over Beeline, the more I think the above feedback is not really applicable here.  Basically old man yelling at cloud situation.
